File: python/examples/example_shamir.py
#!/usr/bin/env python3

"""
Licensed to the Apache Software Foundation (ASF) under one
or more contributor license agreements.  See the NOTICE file
distributed with this work for additional information
regarding copyright ownership.  The ASF licenses this file
to you under the Apache License, Version 2.0 (the
"License"); you may not use this file except in compliance
with the License.  You may obtain a copy of the License at

  http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing,
software distributed under the License is distributed on an
"AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
KIND, either express or implied.  See the License for the
specific language governing permissions and limitations
under the License.
"""

import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))

from amcl import core_utils, shamir

seed_hex = "78d0fb6705ce77dee47d03eb5b9c5d30"

k = 5
n = 10

q = 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141

if __name__ == "__main__":
    # random number generator
    seed = bytes.fromhex(seed_hex)
    rng = core_utils.create_csprng(seed)

    ## Shamir Secret Sharing
    print("Example Shamir Secret Sharing")
    print(f"k = {k}, n = {n}")

    # Generate secret and compute shares
    shares, S = shamir.make_shares(k, n, rng)

    print("\n[Dealer] Generate secret and shares")
    print(f"\tS = {S.hex()}\n")

    for i, (x, y) in enumerate(shares):
        print(f"\tshare[{i}].X = {x.hex()}")
        print(f"\tshare[{i}].Y = {y.hex()}")
        print("")

    # Recover Secret using k of the shares
    S_rec = shamir.recover_secret(shares[:k])

    print("[Player] Recover secret using k shares")
    for i, (x, y) in enumerate(shares[:k]):
        print(f"\tshare[{i}].X = {x.hex()}")
        print(f"\tshare[{i}].Y = {y.hex()}")
        print("")

    print("[Player] Recovered Secret:")
    print(f"\tS = {S_rec.hex()}")

    if S_rec != S:
        print("[Player] Failure Recovering Secret")
        sys.exit(1)

    ## Shamir to Additive Conversion
    #
    # Participants need the X component of the other players shares
    print("\nExample Shamir to Additive Shares conversion")
    print("Using the same shares as above")
    print("\nConvert Shares")
    additive_shares = []
    for i in range (k):
        # Select X component of shares for other participants
        (X, _) = zip(*(shares[:k]))
        X = [x for x in X]
        X.pop(i)

        additive_share = shamir.to_additive(shares[i], X)
        additive_shares.append(additive_share)

    additive_secret = 0
    for i, share in enumerate(additive_shares):
        print(f"\tshare[{i}] = {share.hex()}")
        additive_share = int.from_bytes(share, byteorder='big')
        additive_secret = (additive_secret + additive_share) % q

    additive_secret = additive_secret.to_bytes(shamir.EGS, byteorder='big')
    print("\nReconstructed secret")
    print(f"\tS = {additive_secret.hex()}")

    if additive_secret != S:
        print("[Dealer] Failure Recovering Additive Secret")
        sys.exit(1)

    ## Verifiable Secret Sharing
    print("\nExample Verifiable Secret Sharing")
    print(f"k = {k}, n = {n}")

    # Reuse secret from above, make shares and checks
    shares, checks, _ = shamir.vss_make_shares(k, n, rng, S = S)

    print("\n[Dealer] Make shares and Checks")
    print(f"\tS = {S.hex()}\n")

    for i, (x, y) in enumerate(shares):
        print(f"\tshare[{i}].X = {x.hex()}")
        print(f"\tshare[{i}].Y = {y.hex()}")
        print("")

    for i, c in enumerate(checks):
        print(f"\tcheck[{i}] = {c.hex()}")

    # Verify received share
    for i, share in enumerate(shares):
        rc = shamir.vss_verify_shares(share, checks)
        print(f"\n[Player] {i}-th player - verify received share")
        if rc == shamir.OK:
            print("\tSuccess")
        else:
            print(f"\tFailure: {rc}")
            sys.exit(1)

    # Recover Secret using k of the shares
    S_rec = shamir.recover_secret(shares[:k])

    print("\n[Player] Recover secret using k shares")
    for i, (x, y) in enumerate(shares[:k]):
        print(f"\tshare[{i}].X = {x.hex()}")
        print(f"\tshare[{i}].Y = {y.hex()}")
        print("")

    print("[Player] Recovered Secret:")
    print(f"\tS = {S_rec.hex()}")

    if S_rec != S:
        print("[Player] Failure Recovering Secret")
        sys.exit(1)

File: examples/example_shamir.c
/*
    Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.
*/

/* SSS/VSS example */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "amcl/shamir.h"

int main()
{
    int i;
    int rc;

    int n=4;
    int k=3;

    // Secret
    char s[SGS_SECP256K1];
    octet S = {0,sizeof(s),s};

    char s_golden[SGS_SECP256K1];
    octet S_GOLDEN = {0, sizeof(s_golden), s_golden};

    // Secret shares
    char x[n][SGS_SECP256K1];
    octet X[n];
    char y[n][SGS_SECP256K1];
    octet Y[n];

    for(i = 0; i < n; i++)
    {
        Y[i].max = SGS_SECP256K1;
        Y[i].len = SGS_SECP256K1;
        Y[i].val = y[i];

        X[i].max = SGS_SECP256K1;
        X[i].len = SGS_SECP256K1;
        X[i].val = x[i];
    }

    SSS_shares shares = {X, Y};

    // Deterministic RNG for testing
    char seed[32] = {0};
    csprng RNG;
    RAND_seed(&RNG, 32, seed);

    /* Shamir Secret Sharing */

    printf(" *** Shamir Secret Sharing *** \n");
    printf("\tk = %d, n = %d\n", k, n);

    // Compute shares. The secret is randomly generated here
    SSS_make_shares(k, n, &RNG, &shares, &S_GOLDEN);

    printf("\tS = ");
    OCT_output(&S_GOLDEN);

    printf("\nMake Shares\n");
    for(i = 0; i < n; i++)
    {
        printf("\tX[%d] = ", i);
        OCT_output(X+i);
        printf("\tY[%d] = ", i);
        OCT_output(Y+i);
        printf("\n");
    }

    printf("\nRecover Secret using first k shares\n");

    SSS_recover_secret(k, &shares, &S);

    printf("\tS = ");
    OCT_output(&S);

    if (!OCT_comp(&S, &S_GOLDEN))
    {
        printf("FAILURE - Secrets do not match\n");
        exit(EXIT_FAILURE);
    }

    /* Shamir to additive conversion */

    printf("\n *** Shamir to additive conversion *** \n");
    printf("\tUsing first %d shares from above\n", k);

    char sh[SGS_SECP256K1];
    octet SH = {0, sizeof(sh), sh};

    char others[k-1][SGS_SECP256K1];
    octet OTHERS[k-1];

    for (i = 0; i < k-1; i++)
    {
        OTHERS[i].max = SGS_SECP256K1;
        OTHERS[i].len = SGS_SECP256K1;
        OTHERS[i].val = others[i];
    }

    BIG_256_56 acc;
    BIG_256_56 share;
    BIG_256_56 q;

    BIG_256_56_rcopy(q, CURVE_Order_SECP256K1);
    BIG_256_56_zero(acc);

    printf("\nConverted shares:\n");

    for (i = 0; i < k; i++)
    {
        // Load other X shares into memory
        octet *other_shares_ptr = OTHERS;
        for (int j = 0; j < k; j++)
        {
            if (j == i) continue;

            OCT_copy(other_shares_ptr, X+j);
            other_shares_ptr++;
        }

        SSS_shamir_to_additive(k, X+i, Y+i, OTHERS, &SH);

        printf("\tSH_add[%d] = ", i);
        OCT_output(&SH);

        // Add to accumulator to recover secret using additive shares
        BIG_256_56_fromBytesLen(share, SH.val, SH.len);
        BIG_256_56_add(acc, acc, share);
        BIG_256_56_mod(acc, q);
    }

    BIG_256_56_toBytes(SH.val, acc);
    SH.len = SGS_SECP256K1;

    printf("\nReconstruct (additive) secret\n");
    printf("\tS = ");
    OCT_output(&SH);

    if (!OCT_comp(&SH, &S))
    {
        printf("FAILURE - Secrets do not match\n");
        exit(EXIT_FAILURE);
    }

    /* Verifiable Secret Sharing */
    printf("\n *** Verifiable Secret Sharing *** \n");
    printf("\tk = %d, n = %d\n", k, n);
    printf("\tS = ");
    OCT_output(&S);

    // Additional checks for verification
    char c[k][1 + SFS_SECP256K1];
    octet C[k];

    for(i = 0; i < k; i++)
    {
        C[i].max = 1 + SFS_SECP256K1;
        C[i].len = 1 + SFS_SECP256K1;
        C[i].val = c[i];
    }

    VSS_make_shares(k, n, &RNG, &shares, C, &S_GOLDEN);

    printf("\nMake Shares ...\n");
    for(i = 0; i < n; i++)
    {
        printf("\tX[%d] = ", i);
        OCT_output(X+i);
        printf("\tY[%d] = ", i);
        OCT_output(Y+i);
        printf("\n");
    }

    printf("\n.. and checks\n");
    for(i = 0; i < k; i++)
    {
        printf("\tC[%d] = ", i);
        OCT_output(C+i);
    }

    printf("\nVerify Shares\n");

    for (i = 0; i < n; i++)
    {
        printf("\tShare %d: ", i);
        rc = VSS_verify_shares(k, X+i, Y+i, C);

        if (rc == VSS_OK)
        {
            printf("Success\n");
        }
        else
        {
            printf("Failure\n");
            exit(EXIT_FAILURE);
        }
    }

    printf("\nRecover Secret using first k shares\n");

    SSS_recover_secret(k, &shares, &S);

    printf("\tS = ");
    OCT_output(&S);

    if (!OCT_comp(&S, &S_GOLDEN))
    {
        printf("FAILURE - Secrets do not match\n");
        exit(EXIT_FAILURE);
    }
}


